### PR TITLE
feat(track): Introducing easy event tracking

### PR DESCRIPTION
**Pull request by mjc1283**
## Summary
Updates track method and event builder to support easy event tracking.

## Test plan
- Unit tests: updated, and and added new ones
- Compatibility suite tests
- Manual testing

### CODE REVIEW

File: packages/optimizely-sdk/lib/core/project_config/index.js [hunk 1]
```diff
@@ -605,4 +584,14 @@ module.exports = {
...
     return projectConfig.audiencesById;
   },
+
+  /**
+   * Returns true if an event with the given key exists in the datafile, and false otherwise
+   * @param {Object} projectConfig
+   * @param {string} eventKey
+   * @returns {boolean}
+   */
+  eventWithKeyExists: function(projectConfig, eventKey) {
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/core/project_config/index.js]: This is more of a question. Why did we decide to start passing `projectConfig` to an instance of `projectConfig`.

Is doing `this.eventKeyMap` bad?

Also do you think it makes more sense to use the existing `projectConfig.getEventId` and check the existence vs adding a new API method.

FWIW, I re-wrote projectConfig and use the API method `getEventByKey` and that's enough for event creation

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 2]
```diff
@@ -225,17 +225,11 @@ Optimizely.prototype.track = function(eventKey, userId, attributes, eventTags) {
...
-      // determine which experiments and variations we should be tracking for the given event
-      var validExperimentsToBucketedVariations = this.__getValidExperimentsForEvent(eventKey, userId, attributes);
-      if (!Object.keys(validExperimentsToBucketedVariations).length) {
-        // Return and do not send conversion events if the event is not associated with any running experiments
-        this.logger.log(LOG_LEVEL.WARNING, sprintf(LOG_MESSAGES.EVENT_NOT_ASSOCIATED_WITH_EXPERIMENTS,
-          MODULE_NAME,
-          eventKey));
-        return;
+      if (!projectConfig.eventWithKeyExists(this.configObj, eventKey)) {
+        throw new Error(sprintf(ERROR_MESSAGES.INVALID_EVENT_KEY, MODULE_NAME, eventKey));
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Another question: is giving an invalid `eventKey` actually an error?

File: packages/optimizely-sdk/lib/core/event_builder/index.js [hunk 3]
```diff
@@ -57,6 +57,7 @@ function getCommonEventParams(options) {
     client_name: options.clientEngine,
     client_version: options.clientVersion,
     anonymize_ip: anonymize_ip,
+    enrich_decisions: true,
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/core/event_builder/index.js]: How did we come to the conclusion of using a config in data payload.

I'm guess I'm fine with it, it just seems kind of the wrong place to put it.  This seems like something "meta" to the event, like it should go in a header or query param.  I get that putting it in a header or query param is probably tough given the architecture of `EventDispatcher`.

Wondering the thoughts on this.

File: packages/optimizely-sdk/lib/core/event_builder/index.js [hunk 4]
```diff
@@ -115,26 +116,14 @@ function getImpressionEventParams(configObj, experimentId, variationId) {
...
  * @param  {string} eventKey                  Event key representing the event which needs to be recorded
  * @param  {Object} eventTags                 Values associated with the event.
- * @param  {Array}  experimentsToVariationMap Map of experiment IDs to bucketed variation IDs
  * @param  {Object} logger                    Logger object
  * @return {Object}                           Conversion event params
  */
-function getVisitorSnapshot(configObj, eventKey, eventTags, experimentsToVariationMap, logger) {  
+function getVisitorSnapshot(configObj, eventKey, eventTags, logger) {
   var snapshot = {
-    decisions: [],
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/core/event_builder/index.js]: is it okay to completely remove `decisions` vs keeping it as an empty array.

**Comment by aliabbasrizvi** [comment]
nit. You should update copyright year in every file that you are updating.

File: packages/optimizely-sdk/lib/core/event_builder/index.js (hunk 3 above)
Comment (mikeng13) [reply packages/optimizely-sdk/lib/core/event_builder/index.js]: 👍 I agree header or QS is more appropriate. We'd need the logtier to be able to handle it though. Have we discussed these alternatives with em?

File: packages/optimizely-sdk/lib/core/project_config/index.js (hunk 1 above)
Comment (mikeng13) [reply packages/optimizely-sdk/lib/core/project_config/index.js]: Yeah not married to this pattern. I'd support refactoring this class to house those functions eventually.

File: packages/optimizely-sdk/lib/utils/fns/index.js [hunk 5]
```diff
@@ -34,7 +34,6 @@ module.exports = {
   forEach: require('lodash/forEach'),
   forOwn: require('lodash/forOwn'),
   map: require('lodash/map'),
-  reduce: require('lodash/reduce'),
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/utils/fns/index.js]: Good riddance!

File: packages/optimizely-sdk/lib/core/event_builder/index.js (hunk 3 above)
Comment (mjc1283) [reply packages/optimizely-sdk/lib/core/event_builder/index.js]: @mikeng13 Yea, we have discussed with logtier, and they are on board with this.

**Comment by aliabbasrizvi** [comment]
Triggered tests for this branch: https://travis-ci.com/optimizely/fullstack-sdk-compatibility-suite/builds/97304511
Looks like we need to look at 2 tests.

**Comment by aliabbasrizvi** [comment]
All tests passed: https://travis-ci.com/optimizely/fullstack-sdk-compatibility-suite/builds/97304511
Merging change

**Pull request merged**